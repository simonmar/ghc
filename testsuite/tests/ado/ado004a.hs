{-# LANGUAGE ApplicativeDo, RebindableSyntax #-}
{-# OPTIONS_GHC -ddump-types #-}
module Test where

import Control.Monad
import Prelude

-- This is a do expression that typechecks with only an Applicative constraint
test1 :: Applicative f => (Int -> f Int) -> f Int
test1 f = do
  x <- f 3
  y <- f 4
  return (x + y)

-- Test we can also infer the Applicative version of the type
test2 f = do
  x <- f 3
  y <- f 4
  return (x + y)

-- This one will use join
test3 f g = do
  x <- f 3
  y <- f 4
  g y x

-- This one needs a tuple
test4 f g = do
  x <- f 3
  y <- f 4
  let r = g y x
  r
