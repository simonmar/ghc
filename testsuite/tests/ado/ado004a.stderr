TYPE SIGNATURES
  test1 ::
    forall (f :: * -> *). Applicative f => (Int -> f Int) -> f Int
  test2 ::
    forall a (f :: * -> *) b.
    (Num a, Num b, Applicative f) =>
    (a -> f b) -> f b
  test3 ::
    forall (m :: * -> *) a a1 a2.
    (Monad m, Num a1) =>
    (a1 -> m a2) -> (a2 -> a2 -> m a) -> m a
  test4 ::
    forall (m :: * -> *) b a a1.
    (Monad m, Num a) =>
    (a -> m a1) -> (a1 -> a1 -> m b) -> m b
TYPE CONSTRUCTORS
COERCION AXIOMS
Dependent modules: []
Dependent packages: [base-4.8.0.0, ghc-prim-0.3.1.0,
                     integer-gmp-1.0.0.0]
